### PR TITLE
Assertion Fix in Split Layer

### PR DESCRIPTION
**Pull request by Abdurrahheem**
### Pull Request Readiness Checklist

This PR fixes issue mentioned in [#23663](https://github.com/opencv/opencv/issues/23663) 
Merge with https://github.com/opencv/opencv_extra/pull/1067

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 1]
```diff
@@ -1363,7 +1363,7 @@ void ONNXImporter::parseSplit(LayerParams& layerParams, const opencv_onnx::NodeP
     {
         DictValue splits = layerParams.get("split");
         const int numSplits = splits.size();
-        CV_Assert(numSplits > 1);
+        CV_Assert(numSplits >= 1);
 
         std::vector<int> slicePoints(numSplits - 1, splits.get<int>(0));
```
Comment (dkurt) [inline modules/dnn/src/onnx/onnx_importer.cpp:1368]: How to deal with a zero size vector?

Comment (Abdurrahheem) [reply modules/dnn/src/onnx/onnx_importer.cpp:1368]: Not sure about zero size vectors. But when assertion is  such that `numSplits > 1` OpenCV parser does not pass this construction because of accretion: 

```python 

data  = [torch.ones(N, C, W, H),]

for x in data: 
    x = some processing

```

Comment (dkurt) [reply modules/dnn/src/onnx/onnx_importer.cpp:1368]: I mean the code below which is incorrect for empty vector

Comment (dkurt) [reply modules/dnn/src/onnx/onnx_importer.cpp:1368]: So we should not get pointer of the empty vector:
```cpp
std::vector<int> slicePoints(numSplits - 1, splits.get<int>(0));
layerParams.set("slice_point", DictValue::arrayInt(&slicePoints[0], slicePoints.size()));
```
Can you please create a test? Even without ONNX, just a Split layer initialized manually. See https://github.com/opencv/opencv/blob/4.x/modules/dnn/test/test_layers.cpp for examples.

Comment (Abdurrahheem) [reply modules/dnn/src/onnx/onnx_importer.cpp:1368]: Added test with split = 1

Comment (Abdurrahheem) [reply modules/dnn/src/onnx/onnx_importer.cpp:1368]: All tests seem to be passing. Added test generator is [here](https://github.com/opencv/opencv_extra/pull/1067)

File: modules/dnn/test/test_onnx_importer.cpp [hunk 2]
```diff
@@ -1146,6 +1146,7 @@ TEST_P(Test_ONNX_layers, Split)
         applyTestTag(CV_TEST_TAG_DNN_SKIP_IE_NN_BUILDER);
     if (backend == DNN_BACKEND_INFERENCE_ENGINE_NGRAPH)
         applyTestTag(CV_TEST_TAG_DNN_SKIP_IE_NGRAPH);
+    testONNXModels("split_0");
```
Comment (dkurt) [inline modules/dnn/test/test_onnx_importer.cpp:1149]: This test does not cover code changes. It goes not to `if (layerParams.has("split"))`  branch but `else if (node_proto.input_size() == 2)`.

File: modules/dnn/src/onnx/onnx_importer.cpp (hunk 1 above)
Comment (dkurt) [reply modules/dnn/src/onnx/onnx_importer.cpp:1368]: I'd like suggest the following approach. Having a zero size vector and accessing it's memory is not safe:
```cpp
if (numSplits == 1)
{
    layerParams.set("num_split", 1);
}
else
{
    CV_Assert(numSplits > 1);

    std::vector<int> slicePoints(numSplits - 1, splits.get<int>(0));
    for (int i = 1; i < splits.size() - 1; ++i)
    {
        slicePoints[i] = slicePoints[i - 1] + splits.get<int>(i);
    }
    layerParams.set("slice_point", DictValue::arrayInt(&slicePoints[0], slicePoints.size()));
}
```

**Review by dkurt — APPROVED** [review]
